File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/domain/interactor/location/GetPlacesUseCase.java
package com.lattis.ellipse.domain.interactor.location;

import com.lattis.ellipse.domain.executor.PostExecutionThread;
import com.lattis.ellipse.domain.executor.ThreadExecutor;
import com.lattis.ellipse.domain.interactor.UseCase;
import com.lattis.ellipse.domain.model.Location;
import com.lattis.ellipse.domain.model.map.PlaceAutocomplete;
import com.lattis.ellipse.domain.repository.LocationRepository;

import java.util.ArrayList;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;

public class GetPlacesUseCase extends UseCase<ArrayList<PlaceAutocomplete>> {

    private LocationRepository locationRepository;
    private String constraint;

    @Inject
    protected GetPlacesUseCase(ThreadExecutor threadExecutor,
                                        PostExecutionThread postExecutionThread,
                                        LocationRepository locationRepository) {
        super(threadExecutor, postExecutionThread);
        this.locationRepository = locationRepository;
    }

    public GetPlacesUseCase withConstraint(String constraint){
        this.constraint = constraint;
        return this;
    }

    @Override
    protected Observable<ArrayList<PlaceAutocomplete>> buildUseCaseObservable() {
        return this.locationRepository.getPlaces(constraint);
    }

    @Override
    public Disposable execute(DisposableObserver<ArrayList<PlaceAutocomplete>> UseCaseSubscriber) {
        return this.buildUseCaseObservable()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(UseCaseSubscriber);
    }


}

File: velolabs_repos/lattis_android_v2/domain/src/main/java/com/lattis/domain/usecase/location/GetPlacesUseCase.kt
package com.lattis.domain.usecase.location

import com.lattis.domain.executor.PostExecutionThread
import com.lattis.domain.executor.ThreadExecutor
import com.lattis.domain.models.map.PlaceAutocomplete
import com.lattis.domain.repository.LocationRepository
import com.lattis.domain.usecase.base.UseCase
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.observers.DisposableObserver
import io.reactivex.rxjava3.schedulers.Schedulers
import javax.inject.Inject


class GetPlacesUseCase @Inject
constructor(threadExecutor: ThreadExecutor,
            postExecutionThread: PostExecutionThread,
            private val locationRepository: LocationRepository
) : UseCase<ArrayList<PlaceAutocomplete>>(threadExecutor, postExecutionThread) {
    private lateinit var constraint: String

    fun withConstraint(constraint: String): GetPlacesUseCase {
        this.constraint = constraint
        return this
    }

    override fun buildUseCaseObservable(): Observable<ArrayList<PlaceAutocomplete>> {
        return this.locationRepository.getPlaces(constraint)
    }

    override fun execute(UseCaseSubscriber: DisposableObserver<ArrayList<PlaceAutocomplete>>): Disposable {
        return this.buildUseCaseObservable()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(UseCaseSubscriber)
    }


}
